lte/gateway/c/oai/tasks/amf/amf_app_ue_context_and_proc.h

```diff
@@ -0,0 +1,1001 @@
+#define AMF_HEADER_MINIMUM_LENGTH AMF_HEADER_LENGTH
+#define AMF_HEADER_MAXIMUM_LENGTH AMF_HEADER_LENGTH
+class amf_msg_header {
+ public:
+  uint8_t extended_protocol_discriminator;
+  uint8_t security_header_type;
+  uint8_t message_type;
+  uint8_t sequence_number;
+};
+
+class amf_app_msg {
+ public:
+  void amf_app_ue_context_release(
+      ue_m5gmm_context_s* ue_context_p, ngap_Cause_t cause);
+};
+
+class amf_nas_message_decode_status_t {
+ public:
+  amf_nas_message_decode_status_t() {}
+  ~amf_nas_message_decode_status_t() {}
+  uint8_t integrity_protected_message : 1;
+  uint8_t ciphered_message : 1;
+  uint8_t mac_matched : 1;
+  uint8_t security_context_available : 1;
+  int amf_cause;
+};
+
+class AMFMsg {
+ public:
+  AMFMsg() {}
+  ~AMFMsg() {}
+
+  amf_msg_header header;
+  RegistrationRequestMsg registrationrequestmsg;
+  RegistrationAcceptMsg registrationacceptmsg;
+  RegistrationCompleteMsg registrationcompletemsg;
+  RegistrationRejectMsg registrationrejectmsg;
+  IdentityRequestMsg identityrequestmsg;
+  IdentityResponseMsg identityresponsemsg;
+  AuthenticationRequestMsg authenticationrequestmsg;
+  AuthenticationResponseMsg authenticationresponsemsg;
+  AuthenticationRejectMsg authenticationrejectmsg;
+  AuthenticationFailureMsg authenticationfailuremsg;
+  SecurityModeCommandMsg securitymodecommandmsg;
+  SecurityModeCompleteMsg securitymodecompletemsg;
+  SecurityModeRejectMsg securitymodereject;
+  DeRegistrationRequestUEInitMsg deregistrationequesmsg;
+  DeRegistrationAcceptUEInitMsg deregistrationacceptmsg;
+  ULNASTransportMsg uplinknas5gtransport;
+  DLNASTransportMsg downlinknas5gtransport;
+
+  int amf_msg_decode_header(
+      amf_msg_header* header, const uint8_t* buffer, uint32_t len);
+  int amf_msg_encode_header(
+      const amf_msg_header* header, uint8_t* buffer, uint32_t len);
+  int amf_msg_decode(AMFMsg* msg, uint8_t* buffer, uint32_t len);
+  int AmfMsgEncode(AMFMsg* msg, uint8_t* buffer, uint32_t len);
+};
+
+/* union of plain NAS message */
+typedef struct nas_message_plain_s {
+  AMFMsg amf; /* 5GMM Mobility Management messages */
+} nas_message_plain_t;
+
+typedef struct nas_message_security_protected_s {
+  amf_msg_header header;
+  nas_message_plain_t plain;
+} nas_message_security_protected_t;
+
+typedef struct amf_nas_message_s {
+  amf_msg_header header;
+  nas_message_security_protected_t security_protected;
+  nas_message_plain_t plain;
+} amf_nas_message_t;
+
+typedef enum {
+  CN5G_PROC_NONE = 0,
+  CN5G_PROC_AUTH_INFO,
+} cn5g_proc_type_t;
+
+typedef enum amf_common_proc_type_s {
+  AMF_COMM_PROC_NONE = 0,
+  AMF_COMM_PROC_GUTI,
+  AMF_COMM_PROC_AUTH,
+  AMF_COMM_PROC_SMC,
+  AMF_COMM_PROC_IDENT,
+  AMF_COMM_PROC_INFO,
+} amf_common_proc_type_t;
+
+enum nas_base_proc_type_t {
+  NAS_PROC_TYPE_NONE = 0,
+  NAS_PROC_TYPE_AMF,
+  NAS_PROC_TYPE_SMF,
+  NAS_PROC_TYPE_CN,
+};
+
+class nas5g_base_proc_t;
+class nas_amf_proc_t;
+class nas_amf_registration_proc_t;
```

Comment:
why are these classes?

Reply:
for forward declaration, changed to structs